File: forge_ios/Forge/Proto/AgentProto.swift
import Foundation
import simd

// MARK: - Wire envelope
//
// Mirrors `forge_orchestrator/forge_orchestrator/proto/events.py` and
// `forge_quest/.../proto/AgentProto.kt` 1:1. The JSON discriminator key is
// "kind" and its values are PascalCase ("Transcript", "ToolCall", …) — see
// AgentProtoCoding.swift. FROZEN: do not add or rename cases here; the decoder
// tolerates unknown kinds so a newer orchestrator stays compatible.

enum AgentEvent: Equatable {
    // v1 carryover, with v2-additive fields (specs/00 §2.1).
    case transcript(text: String, partial: Bool, ts: Int64, speaker: Speaker, smeId: String?)
    case toolCall(name: String, argsJson: String, callId: String)
    case toolResult(callId: String, resultJson: String, deferred: Bool)
    case confirmationRequest(callId: String, summary: String, risk: Risk, invokerSmeId: String?, actionCardJson: String?)
    case confirmationResponse(callId: String, approved: Bool, approverChannel: ApproverChannel)
    case audioChunk(pcmBase64: String, ts: Int64)
    case hello(client: String, sessionId: String, protocolVersion: String)
    case goodbye(reason: String)

    // v2 chat bus + control envelopes (specs/00 §2, specs/04).
    case chatMessage(ChatMessage)
    case channelUpdate(ChannelUpdate)
    case channelList(ChannelList)
    case channelHint(ChannelHint)
    case summonGuild(SummonGuild)
    case smeResponse(SmeResponse)
    case dissentReport(DissentReport)
    case safetyInterrupt(SafetyInterrupt)
    case checkpointMarker(CheckpointMarker)
    case replayDone(ReplayDone)
    case backpressureNotice(BackpressureNotice)
    case errorEvent(ErrorEvent)
    case ping(nonce: String)
    case pong(nonce: String)
    case subscribe(channelId: String)
    case unsubscribe(channelId: String)
}

enum Risk: String, Codable, Equatable {
    case low = "LOW"
    case medium = "MEDIUM"
    case high = "HIGH"
}

// MARK: - Camera & audio chunks (binary, sent on a separate WS frame)

struct FrameChunk: Equatable {
    let jpegBytes: Data
    let widthPx: Int
    let heightPx: Int
    let timestampNs: Int64
}

struct AudioInChunk: Equatable {
    let pcm: Data            // 16 kHz mono signed 16-bit little-endian
    let timestampNs: Int64
}

// MARK: - Intrinsics (per-session)

struct CameraIntrinsics: Codable, Equatable {
    let focalLengthPx: SIMD2<Float>     // (fx, fy)
    let principalPointPx: SIMD2<Float>  // (cx, cy)
    let distortionCoeffs: [Float]       // radial k1..k4, or empty if rectified
    let imageSizePx: SIMD2<Int32>       // (width, height)
}

// MARK: - Detections (returned by the orchestrator's look_at_bench tool)

struct Bbox2D: Codable, Equatable, Hashable {
    let x1: Int32
    let y1: Int32
    let x2: Int32
    let y2: Int32

    var centerPx: SIMD2<Float> { SIMD2(Float(x1 + x2) / 2, Float(y1 + y2) / 2) }
    var widthPx: Int32 { x2 - x1 }
    var heightPx: Int32 { y2 - y1 }
}

struct DetectedComponent: Codable, Equatable, Hashable {
    let id: String              // stable per-component identifier (e.g. "U1")
    let partNumber: String      // human-readable label (e.g. "STM32F411")
    let bbox: Bbox2D
    let confidence: Float
    let secondary: String?

    // Phase 2 — present when the orchestrator emits segmentation polygons:
    let maskPolygon: [SIMD2<Float>]?   // image-space contour, normalized 0..1; nil for bbox-only

    init(
        id: String,
        partNumber: String,
        bbox: Bbox2D,
        confidence: Float = 1.0,
        secondary: String? = nil,
        maskPolygon: [SIMD2<Float>]? = nil
    ) {
        self.id = id
        self.partNumber = partNumber
        self.bbox = bbox
        self.confidence = confidence
        self.secondary = secondary
        self.maskPolygon = maskPolygon
    }

    // The orchestrator omits `confidence` (defaults to 1.0) and `maskPolygon`
    // on bbox-only detections; decode tolerantly so those payloads parse.
    private enum CodingKeys: String, CodingKey {
        case id, partNumber, bbox, confidence, secondary, maskPolygon
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        partNumber = try c.decode(String.self, forKey: .partNumber)
        bbox = try c.decode(Bbox2D.self, forKey: .bbox)
        confidence = try c.decodeIfPresent(Float.self, forKey: .confidence) ?? 1.0
        secondary = try c.decodeIfPresent(String.self, forKey: .secondary)
        maskPolygon = try c.decodeIfPresent([SIMD2<Float>].self, forKey: .maskPolygon)
    }
}

// MARK: - Pose

struct Pose6dof: Equatable {
    let positionM: SIMD3<Float>
    let orientationQuat: simd_quatf     // (x, y, z, w)
    let timestampNs: Int64
}

File: quest/app/src/main/java/com/meta/spatial/samples/mixedrealitytemplate/forge/proto/AgentProto.kt
package com.meta.spatial.samples.mixedrealitytemplate.forge.proto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Forge v2 wire protocol — Kotlin client side.
 *
 * Source of truth: `specs/00_wire_protocol.md §2.2` (the `AgentEvent` sealed
 * union) plus `orchestrator/chat_bus/envelopes.py` (the chat-bus-only control
 * messages that are NOT part of the union but ride the same socket). Every type
 * here is validated against the golden corpus under `testdata/wire/`.
 *
 * On the wire the discriminator field is `kind` (PascalCase, == class simple
 * name); all other fields are camelCase. We model every inbound/outbound message
 * as one sealed hierarchy [ForgeMsg] so a single polymorphic decode dispatches
 * on `kind`. Unknown `kind`s throw — the socket layer catches and drops them
 * (forward-compat, WP-3).
 */
@Serializable
sealed class ForgeMsg {

    // ── v1 carryover (unchanged in v2) ──────────────────────────────────────
    @Serializable
    @SerialName("Hello")
    data class Hello(
        val client: String,
        val sessionId: String,
        val protocolVersion: String = "2.0",
    ) : ForgeMsg()

    @Serializable
    @SerialName("Goodbye")
    data class Goodbye(val reason: String) : ForgeMsg()

    @Serializable
    @SerialName("Transcript")
    data class Transcript(
        val text: String,
        val partial: Boolean,
        val ts: Long,
        val speaker: Speaker = Speaker.USER,
        val smeId: String? = null,
    ) : ForgeMsg()

    @Serializable
    @SerialName("ToolCall")
    data class ToolCall(
        val name: String,
        val argsJson: String,
        val callId: String,
    ) : ForgeMsg()

    @Serializable
    @SerialName("ToolResult")
    data class ToolResult(
        val callId: String,
        val resultJson: String,
        val deferred: Boolean = false,
    ) : ForgeMsg()

    @Serializable
    @SerialName("ConfirmationRequest")
    data class ConfirmationRequest(
        val callId: String,
        val summary: String,
        val risk: Risk,
        val invokerSmeId: String? = null,
        val actionCardJson: String? = null,
    ) : ForgeMsg()

    @Serializable
    @SerialName("ConfirmationResponse")
    data class ConfirmationResponse(
        val callId: String,
        val approved: Boolean,
        val approverChannel: ApproverChannel = ApproverChannel.VOICE,
    ) : ForgeMsg()

    @Serializable
    @SerialName("AudioChunk")
    data class AudioChunk(val pcmBase64: String, val ts: Long) : ForgeMsg()

    // ── v2 additions ────────────────────────────────────────────────────────
    @Serializable
    @SerialName("ChatMessage")
    data class ChatMessage(
        val channelId: String,
        val authorId: String,
        val authorKind: AuthorKind,
        val body: String,
        val bodyContentType: BodyContentType = BodyContentType.MARKDOWN,
        val mentions: List<String> = emptyList(),
        val replyToId: String? = null,
        val messageId: String,
        val ts: Long,
        val streaming: Boolean = false,
    ) : ForgeMsg()

    @Serializable
    @SerialName("SummonGuild")
    data class SummonGuild(
        val callId: String,
        val topic: String,
        val smes: List<String>,
        val contextRefs: List<String> = emptyList(),
        val deadlineMs: Int = 30_000,
    ) : ForgeMsg()

    @Serializable
    @SerialName("SmeResponse")
    data class SmeResponse(
        val smeId: String,
        val callId: String,
        val confidence: Float,
        val claim: String,
        val rationale: String,
        val evidence: List<EvidenceRef> = emptyList(),
        val proposedActions: List<ProposedAction> = emptyList(),
        val dissentsWith: List<String> = emptyList(),
        val ts: Long,
    ) : ForgeMsg()

    @Serializable
    @SerialName("DissentReport")
    data class DissentReport(
        val callId: String,
        val parties: List<String>,
        val axis: String,
        val summary: String,
        val pairwise: List<DissentPair>,
        val ts: Long,
    ) : ForgeMsg()

    @Serializable
    @SerialName("ChannelUpdate")
    data class ChannelUpdate(
        val messageId: String,
        val deltaText: String,
        val done: Boolean = false,
        val ts: Long,
    ) : ForgeMsg()

    @Serializable
    @SerialName("SafetyInterrupt")
    data class SafetyInterrupt(
        val severity: Severity,
        val reason: String,
        val suggestedRecoverActions: List<ProposedAction> = emptyList(),
        val ts: Long,
    ) : ForgeMsg()

    @Serializable
    @SerialName("CheckpointMarker")
    data class CheckpointMarker(
        val checkpointId: String,
        val graphNodeName: String,
        val ts: Long,
    ) : ForgeMsg()

    // ── chat-bus-only control envelopes (orchestrator/chat_bus/envelopes.py) ──
    @Serializable
    @SerialName("ChannelList")
    data class ChannelList(val channels: List<ChannelInfo>) : ForgeMsg()

    @Serializable
    @SerialName("Ping")
    data class Ping(val nonce: String) : ForgeMsg()

    @Serializable
    @SerialName("Pong")
    data class Pong(val nonce: String) : ForgeMsg()

    @Serializable
    @SerialName("BackpressureNotice")
    data class BackpressureNotice(val dropped: Int, val sinceTs: Long) : ForgeMsg()

    @Serializable
    @SerialName("ReplayDone")
    data class ReplayDone(val resumeTs: Long, val checkpointId: String? = null) : ForgeMsg()

    @Serializable
    @SerialName("Subscribe")
    data class Subscribe(val channelId: String) : ForgeMsg()

    @Serializable
    @SerialName("Unsubscribe")
    data class Unsubscribe(val channelId: String) : ForgeMsg()

    @Serializable
    @SerialName("ErrorEvent")
    data class ErrorEvent(
        val code: String,
        val message: String,
        val causedByMessageId: String? = null,
        val ts: Long,
    ) : ForgeMsg()

    @Serializable
    @SerialName("ChannelHint")
    data class ChannelHint(
        val channelId: String,
        val hint: String,
        val reason: String,
    ) : ForgeMsg()
}

// ── Enums (lowercase / mime values carry @SerialName to match the wire) ──────
@Serializable
enum class Risk { LOW, MEDIUM, HIGH }

@Serializable
enum class Severity { WARN, HALT }

@Serializable
enum class Speaker {
    @SerialName("user") USER,
    @SerialName("live") LIVE,
    @SerialName("sme") SME,
}

@Serializable
enum class AuthorKind {
    @SerialName("user") USER,
    @SerialName("live") LIVE,
    @SerialName("sme") SME,
    @SerialName("system") SYSTEM,
}

@Serializable
enum class BodyContentType {
    @SerialName("text/markdown") MARKDOWN,
    @SerialName("application/json") JSON,
    @SerialName("text/code") CODE,
}

@Serializable
enum class ApproverChannel {
    @SerialName("voice") VOICE,
    @SerialName("chat") CHAT,
}

@Serializable
enum class Actor {
    @SerialName("operator") OPERATOR,
    @SerialName("guild") GUILD,
}

// ── Supporting types carried inside events (parity with Pydantic §2.1) ───────
@Serializable
data class EvidenceRef(val kind: String, val uri: String, val note: String? = null)

@Serializable
data class ProposedAction(
    val actor: Actor = Actor.OPERATOR,
    val tool: String,
    val argsJson: String,
    val rationale: String,
    val risk: Risk,
    val instruction: String? = null,
    val documentedLimitRef: String? = null,
)

@Serializable
data class DissentPair(
    val a: String,
    val b: String,
    val aClaim: String,
    val bClaim: String,
    val crux: String,
)

@Serializable
data class ChannelInfo(
    val id: String,
    val title: String,
    val smeId: String? = null,
    val icon: String? = null,
    val alwaysVisible: Boolean = false,
    val unreadHint: Int = 0,
)

/**
 * Rendered inside [ForgeMsg.ConfirmationRequest.actionCardJson] — the operator
 * instruction card. Decoded from the nested JSON string with [ForgeJson].
 */
@Serializable
data class ActionCard(
    val title: String,
    val bodyMarkdown: String,
    val diffMarkdown: String? = null,
    val risk: Risk,
    val documentedLimit: String? = null,
    val affirmLabel: String = "I did it",
    val denyLabel: String = "Skip",
)

/** On-demand snapshot result, carried inside a [ForgeMsg.ChatMessage] JSON body. */
@Serializable
data class FrameRef(
    val uri: String,
    val width: Int,
    val height: Int,
    val ts: Long,
    val sourceSeq: Int,
)

@Serializable
data class SnapshotAnalysis(
    val jobId: String,
    val frame: FrameRef,
    val model: String,
    val analysis: String,
    val cites: List<EvidenceRef> = emptyList(),
    val ts: Long,
)

/**
 * The single JSON configured for the Forge wire: `kind` discriminator,
 * tolerant of unknown fields (WP-3) and unknown enum-adjacent extras, and
 * emits defaults so outbound messages carry every required field.
 */
val ForgeJson: Json = Json {
    classDiscriminator = "kind"
    ignoreUnknownKeys = true
    encodeDefaults = true
    explicitNulls = false
}
